chatsky/core/ctx_dict.py

```diff
@@ -0,0 +1,478 @@
+        elif isinstance(other, Mapping):
+            for key in other:
+                self[key] = other[key]
+        elif hasattr(other, "keys"):
+            for key in other.keys():
+                self[key] = other[key]
+        else:
+            for key, value in other:
+                self[key] = value
+        for key, value in kwds.items():
+            self[key] = value
+
+    async def setdefault(self, key: int, default=None) -> BaseModel:
+        try:
+            return await self[key]
+        except KeyError:
+            self[key] = default
+        return default
+
+    def __eq__(self, value: object) -> bool:
+        if isinstance(value, ContextDict):
+            return self._items == value._items
+        elif isinstance(value, Dict):
+            return self._items == value
+        else:
+            return False
+
+    def __repr__(self) -> str:
+        return (
+            f"ContextDict(items={self._items}, "
+            f"keys={list(self.keys())}, "
+            f"hashes={self._hashes}, "
+            f"added={self._added}, "
+            f"removed={self._removed}, "
+            f"storage={self._storage}, "
+            f"ctx_id={self._ctx_id}, "
+            f"field_name={self._field_name})"
+        )
+
+    def __copy__(self):
+        storage = self._storage
+        self._storage = None
+        copy = BaseModel.__copy__(self)
+        copy._storage = self._storage = storage
+        return copy
+
+    def __deepcopy__(self, memo: dict[int, Any] | None = None):
+        storage = self._storage
+        self._storage = None
+        copy = BaseModel.__deepcopy__(self, memo)
+        copy._storage = self._storage = storage
+        return copy
+
+    @model_validator(mode="wrap")
+    def _validate_model(value: Any, handler: Callable[[Any], "ContextDict"], _) -> "ContextDict":
+        if isinstance(value, ContextDict):
+            return value
+        elif isinstance(value, Dict):
+            instance = handler(dict())
+            instance._items = value.copy()
+            instance._keys = set(value.keys())
+            return instance
+        else:
+            raise ValueError(f"Unknown type of ContextDict value: {type(value).__name__}!")
+
+    def _serialize_model_base(self, to_bytes: bool = False) -> Dict[int, Union[BaseModel, bytes]]:
+        if self._storage is None:
+            return self._items
+        elif not self._storage.rewrite_existing:
+            result = dict()
+            for k, v in self._items.items():
+                value = self._value_type.dump_json(v)
+                if _get_hash(value) != self._hashes.get(k, None):
+                    result[k] = value if to_bytes else v
+            return result
+        else:
+            return {k: self._value_type.dump_json(self._items[k]) if to_bytes else self._items[k] for k in self._added}
+
+    @model_serializer()
+    def _serialize_model(self) -> Dict[int, BaseModel]:
+        return self._serialize_model_base()
```

Comment:
Why make a `model_serializer` out of this function?
The `store` method only uses `_serialize_model_base`.
`model_dump` for this class is only called in tests.

It doesn't make much sense why serialized `ContextDict` is only the items that need to be updated.
Seems better to leave only `_serialize_model_base` and rename it to something like `_get_updates`.